colossalai/fx/tracer/tracer.py

```diff
@@ -0,0 +1,305 @@
+#!/usr/bin/env python
+"""
+tracer.py: 
+    Implemented a tracer which supports control flow and user-defined meta arguments.
+    The implementation is partly inspired HuggingFace's fx tracer
+"""
+
+import inspect
+import math
+import functools
+import torch
+import torch.nn as nn
+from torch import Tensor
+from torch.fx import Tracer
+from torch.fx.graph import Graph
+from torch.fx.proxy import Proxy, ParameterProxy
+from torch.utils import _pytree
+from ..proxy import ColoProxy
+from typing import Optional, Dict, Any
+from ._tracer_utils import is_element_in_list, extract_meta
+from .meta_patch import meta_patched_function, meta_patched_module
+
+__all__ = ['ColoTracer']
+
+
+class ColoTracer(Tracer):
+    """
+    ColoTracer is a symbolic tracer designed to support dynamic control flow by using meta tensors for the `colossalai.fx` module.
+    This tracer is initialized in the same way as the original torch.fx.Tracer.
+
+    Usage:
+        class Model(nn.Module):
+            def __init__(self):
+                super().__init__()
+                self.linear1 = nn.Linear(10, 10)
+                self.linear2 = nn.Linear(10, 10)
+
+            def forward(self, x, y):
+                x1 = self.linear1(x)
+                y1 = self.linear2(y)
+
+                if x1.dim() == 2:
+                    return x1 + y1
+                else:
+                    return x1 - y1
+
+        model = Model()
+        tracer = ColoTracer()
+        graph = tracer.trace(model, concrete_args={'y': torch.rand(4, 10)}, meta_args={'x': torch.rand(4, 10, device='meta')})
+    """
+
+    # Feature flag for proxying accesses to buffer values
+    proxy_buffer_attributes: bool = True
+
+    _TORCH_METHODS_TO_PATCH = ["arange", "zeros", "ones", "full", "full_like", "eye", "empty", "tensor"]
+
+    def create_proxy(self, kind, target, args, kwargs, name=None, type_expr=None, proxy_factory_fn=None) -> ColoProxy:
```

Comment:
why not *args, **kwargs,

Reply:
These arguments are guaranteed to be backward compatible by PyTorch and some of them are needed for meta-tensor-related logic.